### ISSUE TITLE
2.3 control flow . cell 8 . the exercise of Polynomial Circuit

### ISSUE THREAD

**Issue by NeutrinoLiu** [ucb-bar/generator-bootcamp#33]
the program of generating the circuit didn't work even if I brought in correct solution into "???". and the error info is:

cmd8.sc:29: overloaded method value expect with alternatives:
  (signal: chisel3.Bundle,expected: Map[String,BigInt])Boolean <and>
  (signal: chisel3.core.Aggregate,expected: IndexedSeq[BigInt])Boolean <and>
  (signal: chisel3.Bits,expected: BigInt,msg: => String)Boolean <and>
  (good: Boolean,msg: => String)Boolean
 cannot be applied to (chisel3.core.SInt, AnyVal)
      expect(c.io.fOfX, poly(select, x))
      ^
Compilation Failed

**Comment by edwardcwang** [comment]
It seems like the return type of your `poly()` is `AnyVal`, which suggests a possible bug in the `poly()` function.

**Comment by NeutrinoLiu** [comment]
yes that makes sense, Ive found the problem: in the former scala numeric exercise, I used **"if(select==0)else if(select==1)else if(select==2)"**  instead of **"if(select==0)else if(select==1)else ..."** to do the conditional branch. it works now.  but it seems not a problem of the return type of poly() ? To be honest I suppose **"if(select==0)else if(select==1)else if(select==2)"** and **"if(select==0)else if(select==1)else ..."** have almost no difference in this case, isn't it? even it fixes the problem, I still don't know why...moreover, in the former use of the function **expect**, expect(IO, val) is often the case.

**Comment by edwardcwang** [comment]
If you define a function that returns in some branches, not others, Scala will infer the result type to be `AnyVal`. For example:

```scala
def funcTest(sel: Int) = {
    if (sel == 5) {
        10
    }
```

is equivalent to:


```scala
def funcTest(sel: Int): AnyVal = {
    if (sel == 5) {
        10
    }
}
```

**Comment by NeutrinoLiu** [comment]
emm, I see. so that means:
a) the type  `AnyVal != Val` 
b) if the conditions listed are not complete the function will be defined as AnyVal defaultly, instead of a normal Val.
correct?

**Comment by edwardcwang** [comment]
a) `Val` is not a Scala type and lowercase `val` is a keyword and cannot be used as a variable name
b) The return type if none of the if conditions are triggered is `Unit`, and Scala will determine the result type to be the common supertype of the branch (`Int`) and `Unit`, which is `AnyVal`. Note that `AnyVal` is much less useful and should generally be avoided. If you need to sometimes return a type, Scala has the [Option type](https://danielwestheide.com/blog/2012/12/19/the-neophytes-guide-to-scala-part-5-the-option-type.html).

In other words:

```scala
def funcTest(sel: Int): AnyVal = {
    if (sel == 5) {
        10
    }
}
```

is identical to:

```scala
def funcTest(sel: Int): AnyVal = {
    if (sel == 5) {
        10
    } else {
        Unit
    }
}
```

**Comment by NeutrinoLiu** [comment]
I believe this time I get it~thats really helpful! 
thanks for your patience!

**Comment by edwardcwang** [comment]
Glad to hear it!
